### PR TITLE
feature(735-handle-missing-federation-member-pk): handling  federation toml missing public keys

### PR DESCRIPTION
**Pull request by nagarajm87**
* handling  `federation.toml` missing public keys.

### CODE REVIEW

File: bin/reth/src/commands/poa/mod.rs [hunk 1]
```diff
@@ -621,26 +621,34 @@ impl<Ext: clap::Args + fmt::Debug> PoaNodeCommand<Ext> {
...
-            let authority_index = genesis_authorities
-                .iter()
-                .position(|a| a == &authority_pk)
-                .expect("Authority not found in genesis authorities");
+            let authority_index = match genesis_authorities.iter().position(|a| a == &authority_pk)
+            {
+                Some(index) => index,
+                None => {
+                    error!("Federation public keys not found ");
+                    Default::default()
```
Comment (arminsabouri) [inline bin/reth/src/commands/poa/mod.rs]: Should we error out of the program here? If a node is saying they are a fed node but their key is not found maybe we should just exit

**Review by arminsabouri — CHANGES_REQUESTED** [review]

File: bin/reth/src/commands/poa/mod.rs [hunk 2]
```diff
@@ -781,10 +787,15 @@ impl<Ext: clap::Args + fmt::Debug> PoaNodeCommand<Ext> {
...
             RandomSourceProvider::new(),
             canon_state_notification_receiver,
-        )
-        .expect("Failed to create authority consensus builder")
-        .build()
-        .await;
+        ) {
+            Ok(consensus) => consensus.build().await,
+            Err(e) => {
+                return Err(eyre::eyre!(
```
Comment (scottmillner) [inline bin/reth/src/commands/poa/mod.rs]: `try_new` can fail for other reasons besides `missing federation public keys`

File: crates/node/core/src/args/federation_args.rs [hunk 3]
```diff
@@ -84,19 +88,22 @@ impl FederationTomlConfig {
     ) -> Result<Vec<(secp256k1::PublicKey, SocketAddr)>, Error> {
         let federation_members = self
             .federation_member_public_key
-            .iter()
-            .map(|key| {
-                let public_key = secp256k1::PublicKey::from_str(&key.key)
-                    .expect("Invalid hex string for PublicKey");
+            .as_ref() // Get a reference to the inner Vec if it exists
```
Comment (rwlockbg) [inline crates/node/core/src/args/federation_args.rs]: Can you remove this comment please ?

File: crates/node/core/src/args/federation_args.rs [hunk 4]
```diff
@@ -84,19 +88,22 @@ impl FederationTomlConfig {
...
+            .map_or(Vec::new(), |members| {
+                members
+                    .iter()
+                    .map(|member| {
+                        let public_key = secp256k1::PublicKey::from_str(&member.key)
+                            .expect("Invalid hex string for PublicKey");
```
Comment (rwlockbg) [inline crates/node/core/src/args/federation_args.rs]: Can we not expect here, but rather collect all Result<PubKey, Error> and check i any are Err typed and then create aan Error message just to know the exact reason ?

File: crates/node/core/src/args/federation_args.rs [hunk 5]
```diff
@@ -84,19 +88,22 @@ impl FederationTomlConfig {
...
 
-                let soc_addr = key.socket_addr.parse::<SocketAddr>().unwrap();
-                (public_key, soc_addr)
-            })
-            .collect::<Vec<(secp256k1::PublicKey, SocketAddr)>>();
+                        let soc_addr = member.socket_addr.parse::<SocketAddr>().unwrap();
```
Comment (rwlockbg) [inline crates/node/core/src/args/federation_args.rs]: same here - to avoid the unwraps somwhow

File: bin/reth/src/commands/poa/mod.rs (hunk 2 above)
Comment (rwlockbg) [reply bin/reth/src/commands/poa/mod.rs]: agree, maybe you can check the err type (if possible) and then react to it ?

Comment (arminsabouri) [reply bin/reth/src/commands/poa/mod.rs]: +1

File: crates/node/core/src/args/federation_args.rs [hunk 6]
```diff
@@ -45,7 +45,7 @@ pub struct FedMemberPubKey {
 #[serde(deny_unknown_fields, rename_all = "kebab-case")]
 pub struct FederationTomlConfig {
     /// federation members public keys
-    pub federation_member_public_key: Vec<FedMemberPubKey>,
+    pub federation_member_public_key: Option<Vec<FedMemberPubKey>>,
```
Comment (arminsabouri) [inline crates/node/core/src/args/federation_args.rs]: I dont think this should ever be optional

**Review by arminsabouri — CHANGES_REQUESTED** [review]

**Review by rwlockbg — DISMISSED** [review]

File: bin/reth/src/commands/poa/mod.rs [hunk 7]
```diff
@@ -621,26 +621,32 @@ impl<Ext: clap::Args + fmt::Debug> PoaNodeCommand<Ext> {
...
         // create frost config if in federation mode
         let frost_config = if is_fed_node {
-            let authority_index = genesis_authorities
-                .iter()
-                .position(|a| a == &authority_pk)
-                .expect("Authority not found in genesis authorities");
+            let authority_index =
+                genesis_authorities.iter().position(|a| a == &authority_pk).ok_or_else(|| {
+                    error!("Federation public keys not found");
+                    eyre::eyre!("Federation public keys not found")
```
Comment (arminsabouri) [inline bin/reth/src/commands/poa/mod.rs]: key* singular not plural. 

Maybe a better error message is "Your public key could not be found in the list of federation public keys"

**Review by arminsabouri — CHANGES_REQUESTED** [review]

**Review by arminsabouri — APPROVED** [review]

**Pull request merged**